playbooks/app_dev/APM_guidance.adoc

```diff
@@ -0,0 +1,156 @@
+
+APM is nowadays achieved by instrumenting the application so that it can generate metrics about what is going on inside the process running it and collecting those metrics to some central database that can then be used to generate dashboards and do alerting.
+
+image::images/APMGuidance-APMArchitecture.png[]
+
+Normally an application is instrumented by adding a probe that captures the main application events and metrics. In Java this is done via an agent (activated usually with the https://docs.oracle.com/javase/8/docs/technotes/tools/windows/java.html#BABDJJFI[CLI argument ]). Other runtimes may have different approaches.
+
+In the container world there are two ways of installing APM tools.
+
+=== Agent included in the container image
+
+In this architecture the container image of the application that needs to be monitored has the agent onboard. Upon start of the container, the agent will be activated and start communicating with the metrics collector.
+
+image::images/APMGuidance-ContainerAPMArch1.png[]
+
+Notice that this approach requires specifically crafted images and decreases the ability to reuse externally prepared images such as the https://docs.openshift.com/container-platform/latest/creating_images/s2i.html[s2i images] from Red Hat. 
+
+==== Prometheus instrumentation
+
+https://prometheus.io/[Prometheus] is becoming a very popular metrics collector and going forward will be officially supported in OpenShift for collecting infrastructure-related metrics. A possible approach to APM is to have Prometheus also collect application metrics. http://www.opensourcerers.org/monitoring-camel-prometheus-red-hat-openshift/[Here] is an example that uses a JMX exporter to do it.
+
+https://projects.spring.io/spring-boot/[Spring Boot] can expose metrics directly via the https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#production-ready[Actuator library]. Integrations between Actuator and Prometheus exist in the open source space, for example https://github.com/moelholm/prometheus-spring-boot-starter[here].
+
+Prometheus has the ability to automatically start scraping pods with some specific annotations.  +
+So the setup for this configuration is very minimal. On the other hand, Prometheus does not currently offer a APM-specific dashboard. 
+
+=== Agent deployed as a daemon set
+
+In this approach the agents are deployed as https://docs.openshift.com/container-platform/latest/dev_guide/daemonsets.html[daemon sets]. The containers in this daemon set are privileged and can inspect the other containers running on the same node. Once a new container starts, the agent sends all available information of that container to the metrics collector.
+
+image::images/APMGuidance-ContainerAPMArch2.png[]
+
+Note that in this case very little additional work is required in order to enable APM and that no additional image configuration is needed.
+
+=== Market players
+
+
+Major market players in APM are:
```

Comment:
Mention OpenShift Primed here:

https://hub.openshift.com/primed

Reply:
Can you explain where exactly you would like to see this link and why?

Reply:
Something like

[OpenShift Primed] lists several market players.  Below are some links to their APM offerings for OpenShift.

Reply:
added a consideration on primed